data-serving/scripts/convert-data/geocode_util.py

```diff
@@ -0,0 +1,200 @@
+from typing import Any, Callable
+from constants import COMMON_LOCATION_ABBREVIATIONS
+
+
+def lookup_location(geocoder: Any, location_tokens: [str]) -> Any:
+    '''
+    Attempts to match the provided location tokens against known locations using
+    the provided geocoder.
+
+    Parameters:
+        location_tokens: One or more tokens representing a location, which 
+          may represent a part or whole of a location of any resolution.
+          Examples: ['NY'], ['Paris, France'], ['Wuhan City, Hubei, China'],
+          ['China']
+
+    Raises:
+        ValueError: When the location consists of more than 3 tokens.
+    Returns:
+      None: When no location tokens are passed in or no geocoding match is
+        found.
+      [Dict[str, Any]]: Location data of the form:
+        {
+          lat: float
+          lng: float
+          geo_resolution: str
+          country_new: str
+          admin_id: int
+          location: str
+          admin3: str
+          admin2: str
+          admin1: str
+        }
+    '''
+    normalized_tokens = [COMMON_LOCATION_ABBREVIATIONS[l.lower()]
+                         if l.lower() in COMMON_LOCATION_ABBREVIATIONS else l
+                         for l in location_tokens]
+
+    if normalized_tokens == []:
+        return None
+    if len(normalized_tokens) == 1:
+        # A location of a single token, which may be a country, city, county,
+        # etc. Ex. "Paris", "China", "FL"
+        return lookup_single_part_location(geocoder, normalized_tokens[0])
+    if len(normalized_tokens) == 2:
+        # A two-part location, which may be (county, country), (city, county),
+        # or (city, country) pairs. Ex. "Paris, France", "Hubei, China",
+        # "Boston, MA"
+        return lookup_two_part_location(
+            geocoder, normalized_tokens[0],
+            normalized_tokens[1])
+    if len(normalized_tokens) == 3:
+        # A three-part location, presumed to including city, county, and
+        # country.
+        return geocoder.geocode(
+            normalized_tokens[0],  # city
+            normalized_tokens[1],  # county
+            normalized_tokens[2])  # country
+
+    raise ValueError('Too many tokens in location')
```

Comment:
Another nit: this has the feeling of being a precondition (i.e. caller has to get this right or they won't go into space today), so I'd put it at the top of the function before any other statements like normalizing the tokens. Partly because we don't need to do that work if the input is wrong, partly so that the whole method body (as it stands now, and as it may change later) can work with the assumption that the input is valid.

Reply:
good point, done!